io/zenoh-transport/src/common/pipeline.rs

```diff
@@ -16,427 +16,182 @@ use super::conduit::{TransportChannelTx, TransportConduitTx};
-        zserialize!();
-
-        // The first serialization attempt has failed. This means that the current
-        // batch is full. Therefore:
-        //   1) remove the current batch from the IN pipeline
-        //   2) add the batch to the OUT pipeline
-        if let Some(batch) = in_guard.try_pull() {
-            // The previous batch wasn't empty
-            let mut out_guard = zlock!(self.stage_out);
-            out_guard[priority].push(batch);
-            drop(out_guard);
-            self.cond_canpull.notify_one();
-
-            // Attempt the serialization on a new empty batch
-            zserialize!();
-        }
-
-        log::warn!(
-            "Transport message dropped because it can not be fragmented: {:?}",
-            message
-        );
-
-        false
-    }
-
-    #[inline]
-    pub(crate) fn push_zenoh_message(&self, mut message: ZenohMessage) -> bool {
-        // If the queue is not QoS, it means that we only have one priority with index 0.
-        let priority = if self.is_qos() {
-            message.channel.priority as usize
-        } else {
-            message.channel.priority = Priority::default();
-            0
-        };
-        // Lock the channel. We are the only one that will be writing on it.
-        let mut ch_guard = if message.is_reliable() {
-            zlock!(self.conduit[priority].reliable)
-        } else {
-            zlock!(self.conduit[priority].best_effort)
-        };
-        // Lock the stage in containing the serialization batches.
-        let mut in_guard = zlock!(self.stage_in[priority]);
-
         macro_rules! zserialize {
-            () => {
-                // Get the current serialization batch. Drop the message
-                // if no batches are available
-                let batch = zgetbatch!(self, priority, in_guard, message.is_droppable());
-                let mp = message.channel.priority;
-                if batch.serialize_zenoh_message(&mut message, mp, &mut ch_guard.sn) {
-                    self.bytes_in[priority].store(batch.len(), Ordering::Release);
-                    self.cond_canpull.notify_one();
+            ($batch:expr) => {{
+                if $batch.serialize_zenoh_message(&mut msg, priority, &mut channel.sn) {
+                    if !self.batching {
+                        drop(c_guard);
+                        self.s_out.move_batch($batch);
+                    } else {
+                        let bytes = $batch.len();
+                        *c_guard = Some($batch);
+                        drop(c_guard);
+                        self.s_out.notify(bytes);
+                    }
                     return true;
                 }
-            };
+            }};
         }
 
+        // Get the current serialization batch.
+        let mut batch = zgetbatch!(false);
         // Attempt the serialization on the current batch
-        zserialize!();
+        zserialize!(batch);
```

Comment:
we should mark macros that early-return somehow (`<macro>_rets` convention?), to improve readability

Reply:
Fixed